File: app/src/main/java/com/meteomontana/android/ui/screens/detail/BoulderDraftStore.kt
package com.meteomontana.android.ui.screens.detail

import android.content.Context
import androidx.compose.ui.geometry.Offset
import org.json.JSONArray
import org.json.JSONObject

/**
 * Piedras a medias: lo que llevabas escrito cuando cerraste el formulario.
 *
 * Proponer una piedra es largo —nombre, modalidad, orientación, una foto por
 * cara y una línea dibujada por vía—, y hasta ahora cerrar por error, recibir
 * una llamada o quedarte sin batería significaba empezar de cero. Eso hace que
 * la gente no lo intente en el sitio, que es justo cuando tiene la roca
 * delante.
 *
 * Se guarda **solo en este móvil**, nunca en el servidor: es tuyo, está a
 * medias y no tiene por qué verlo un admin hasta que lo envíes.
 *
 * Un borrador por escuela: lo normal es abrir varias piedras en la misma
 * jornada y en el mismo sitio, y así el aviso de "tienes una a medias" aparece
 * donde tiene sentido.
 *
 * No se inyecta ni vive en el ViewModel: es estado de una pantalla que solo
 * necesita `Context`. Meterlo en el ViewModel obligaba a que sus tests
 * fabricaran uno —y los rompio— por algo que no tiene nada que ver con ellos.
 *
 * LÍMITE CONOCIDO: las fotos se guardan como `Uri`, que es un puntero a la
 * galería, no una copia. Si el usuario borra la foto del móvil, el borrador
 * conserva el texto y las líneas pero pierde la imagen. Copiarlas a la app
 * costaría espacio en disco por algo que casi siempre se termina el mismo día;
 * si algún día molesta, este es el sitio donde cambiarlo.
 */
class BoulderDraftStore(private val context: Context) {

    private val prefs by lazy {
        context.getSharedPreferences("cumbre_borradores_piedra", Context.MODE_PRIVATE)
    }

    /** Todo lo que hace falta para dejar el formulario como estaba. */
    data class Draft(
        val schoolId: String,
        val lat: Double,
        val lon: Double,
        val name: String,
        val discipline: String,
        val geometry: String,
        val direction: String,
        val sectorBlockId: String?,
        val orientation: String?,
        val path: List<Pair<Double, Double>>,
        val faces: List<BoulderFaceForm>,
        /** Cuándo se guardó, para poder decir "hace 2 días". */
        val savedAt: Long
    )

    fun save(draft: Draft) {
        prefs.edit().putString(draft.schoolId, aJson(draft).toString()).apply()
    }

    fun load(schoolId: String): Draft? =
        prefs.getString(schoolId, null)?.let {
            runCatching { deJson(JSONObject(it)) }.getOrNull()
        }

    fun clear(schoolId: String) {
        prefs.edit().remove(schoolId).apply()
    }

    // ── Serialización ───────────────────────────────────────────────────────
    // A mano y no con una librería porque el formulario es estado de UI, no un
    // DTO del servidor: si mañana cambia un campo, lo que NO puede pasar es que
    // reviente al leer un borrador viejo. De ahí que `load` devuelva null ante
    // cualquier problema en vez de propagar el error.

    private fun aJson(d: Draft) = JSONObject().apply {
        put("schoolId", d.schoolId)
        put("lat", d.lat); put("lon", d.lon)
        put("name", d.name)
        put("discipline", d.discipline)
        put("geometry", d.geometry)
        put("direction", d.direction)
        put("sectorBlockId", d.sectorBlockId ?: JSONObject.NULL)
        put("orientation", d.orientation ?: JSONObject.NULL)
        put("savedAt", d.savedAt)
        put("path", JSONArray().apply {
            d.path.forEach { put(JSONArray().put(it.first).put(it.second)) }
        })
        put("faces", JSONArray().apply {
            d.faces.forEach { cara ->
                put(JSONObject().apply {
                    put("photoUri", cara.photoUri?.toString() ?: JSONObject.NULL)
                    put("orientation", cara.orientation ?: JSONObject.NULL)
                    put("bloques", JSONArray().apply {
                        cara.bloques.forEach { b ->
                            put(JSONObject().apply {
                                put("name", b.name)
                                put("grade", b.grade ?: JSONObject.NULL)
                                put("startType", b.startType ?: JSONObject.NULL)
                                put("description", b.description ?: JSONObject.NULL)
                                put("variant", b.variant ?: JSONObject.NULL)
                                put("linePath", JSONArray().apply {
                                    b.linePath.forEach { p -> put(JSONArray().put(p.x).put(p.y)) }
                                })
                            })
                        }
                    })
                })
            }
        })
    }

    private fun deJson(o: JSONObject): Draft {
        val path = o.optJSONArray("path")?.let { arr ->
            (0 until arr.length()).map {
                val p = arr.getJSONArray(it); p.getDouble(0) to p.getDouble(1)
            }
        } ?: emptyList()

        val caras = o.optJSONArray("faces")?.let { arr ->
            (0 until arr.length()).map { i ->
                val c = arr.getJSONObject(i)
                BoulderFaceForm(
                    photoUri = c.optString("photoUri", "").takeIf { it.isNotBlank() }
                        ?.let { android.net.Uri.parse(it) },
                    orientation = c.optString("orientation", "").takeIf { it.isNotBlank() },
                    bloques = c.optJSONArray("bloques")?.let { bs ->
                        (0 until bs.length()).map { j ->
                            val b = bs.getJSONObject(j)
                            BoulderBloqueForm(
                                name = b.optString("name", ""),
                                grade = b.optString("grade", "").takeIf { it.isNotBlank() },
                                startType = b.optString("startType", "").takeIf { it.isNotBlank() },
                                description = b.optString("description", "").takeIf { it.isNotBlank() },
                                variant = b.optString("variant", "").takeIf { it.isNotBlank() },
                                linePath = b.optJSONArray("linePath")?.let { ps ->
                                    (0 until ps.length()).map { k ->
                                        val p = ps.getJSONArray(k)
                                        Offset(p.getDouble(0).toFloat(), p.getDouble(1).toFloat())
                                    }
                                } ?: emptyList()
                            )
                        }
                    } ?: listOf(BoulderBloqueForm())
                )
            }
        } ?: listOf(BoulderFaceForm())

        return Draft(
            schoolId = o.getString("schoolId"),
            lat = o.getDouble("lat"), lon = o.getDouble("lon"),
            name = o.optString("name", ""),
            discipline = o.optString("discipline", "BOULDER"),
            geometry = o.optString("geometry", "POINT"),
            direction = o.optString("direction", "LTR"),
            sectorBlockId = o.optString("sectorBlockId", "").takeIf { it.isNotBlank() },
            orientation = o.optString("orientation", "").takeIf { it.isNotBlank() },
            path = path,
            faces = caras,
            savedAt = o.optLong("savedAt", 0L)
        )
    }
}

/** ¿Tiene el borrador algo que merezca la pena guardar? */
fun BoulderDraftStore.Draft.tieneContenido(): Boolean =
    name.isNotBlank() ||
        faces.any { cara ->
            cara.photoUri != null ||
                cara.bloques.any { it.name.isNotBlank() || it.grade != null || it.linePath.isNotEmpty() }
        }

File: iosApp/iosApp/Screens/BoulderDraftStore.swift
import Foundation
import UIKit

/// Piedras a medias: lo que llevabas escrito cuando cerraste el formulario.
/// Espejo de `BoulderDraftStore.kt` de Android.
///
/// Proponer una piedra es largo —nombre, modalidad, orientación, una foto por
/// cara y una línea dibujada por vía—, y hasta ahora cerrar por error, recibir
/// una llamada o quedarte sin batería significaba empezar de cero. Eso hace que
/// la gente no lo intente en el sitio, que es justo cuando tiene la roca
/// delante.
///
/// Se guarda **solo en este móvil**, nunca en el servidor: es tuyo, está a
/// medias y no tiene por qué verlo un admin hasta que lo envíes.
///
/// DIFERENCIA CON ANDROID: allí la foto es una referencia a la galería y basta
/// con guardar su dirección. Aquí la foto vive en memoria, así que hay que
/// escribirla a disco (JPEG al 85% en la carpeta de la app) o se pierde al
/// cerrar. A cambio, el borrador de iOS sobrevive aunque borres la foto del
/// carrete.
enum BoulderDraftStore {

    struct Draft {
        var schoolId: String
        var lat: Double
        var lon: Double
        var name: String
        var discipline: String
        var geometry: String
        var direction: String
        var sectorId: String?
        var orientation: String?
        var path: [[Double]]
        var faces: [BoulderFaceForm]
        var savedAt: Double
    }

    private static var carpeta: URL {
        let base = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("borradores-piedra", isDirectory: true)
        try? FileManager.default.createDirectory(at: base, withIntermediateDirectories: true)
        return base
    }

    private static func ficheroJson(_ schoolId: String) -> URL {
        carpeta.appendingPathComponent(seguro(schoolId) + ".json")
    }

    /// El id de escuela va en el nombre del fichero: fuera todo lo que no sea
    /// letra, número o guion, no vaya a colarse una barra y escribir donde no toca.
    private static func seguro(_ s: String) -> String {
        String(s.map { $0.isLetter || $0.isNumber || $0 == "-" ? $0 : "_" })
    }

    // MARK: - Guardar / cargar / borrar

    static func save(_ d: Draft) {
        var caras: [[String: Any]] = []
        for (i, cara) in d.faces.enumerated() {
            var dicc: [String: Any] = [:]
            if let img = cara.photo, let data = img.jpegData(compressionQuality: 0.85) {
                let nombre = seguro(d.schoolId) + "-cara\(i).jpg"
                try? data.write(to: carpeta.appendingPathComponent(nombre))
                dicc["photoFile"] = nombre
            }
            if let o = cara.orientation { dicc["orientation"] = o }
            dicc["blocks"] = cara.blocks.map { b -> [String: Any] in
                var v: [String: Any] = ["name": b.name]
                if let g = b.grade { v["grade"] = g }
                if let s = b.startType { v["startType"] = s }
                if !b.descriptionText.isEmpty { v["descriptionText"] = b.descriptionText }
                if !b.variant.isEmpty { v["variant"] = b.variant }
                v["line"] = b.line.map { [Double($0.x), Double($0.y)] }
                return v
            }
            caras.append(dicc)
        }
        let raiz: [String: Any] = [
            "schoolId": d.schoolId, "lat": d.lat, "lon": d.lon,
            "name": d.name, "discipline": d.discipline,
            "geometry": d.geometry, "direction": d.direction,
            "sectorId": d.sectorId as Any, "orientation": d.orientation as Any,
            "path": d.path, "faces": caras, "savedAt": d.savedAt
        ]
        if let data = try? JSONSerialization.data(withJSONObject: raiz) {
            try? data.write(to: ficheroJson(d.schoolId))
        }
    }

    /// Devuelve nil ante CUALQUIER problema: un borrador viejo con un campo que
    /// ya no existe no puede impedir proponer una piedra nueva.
    static func load(schoolId: String) -> Draft? {
        guard let data = try? Data(contentsOf: ficheroJson(schoolId)),
              let o = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
        else { return nil }

        var caras: [BoulderFaceForm] = []
        for c in (o["faces"] as? [[String: Any]] ?? []) {
            var cara = BoulderFaceForm()
            if let f = c["photoFile"] as? String,
               let d = try? Data(contentsOf: carpeta.appendingPathComponent(f)) {
                cara.photo = UIImage(data: d)
            }
            cara.orientation = c["orientation"] as? String
            cara.blocks = (c["blocks"] as? [[String: Any]] ?? []).map { b in
                var v = BoulderBlockForm()
                v.name = b["name"] as? String ?? ""
                v.grade = b["grade"] as? String
                v.startType = b["startType"] as? String
                v.descriptionText = b["descriptionText"] as? String ?? ""
                v.variant = b["variant"] as? String ?? ""
                v.line = (b["line"] as? [[Double]] ?? []).map {
                    CGPoint(x: $0.first ?? 0, y: $0.count > 1 ? $0[1] : 0)
                }
                return v
            }
            if cara.blocks.isEmpty { cara.blocks = [BoulderBlockForm()] }
            caras.append(cara)
        }
        if caras.isEmpty { caras = [BoulderFaceForm()] }

        return Draft(
            schoolId: o["schoolId"] as? String ?? schoolId,
            lat: o["lat"] as? Double ?? 0, lon: o["lon"] as? Double ?? 0,
            name: o["name"] as? String ?? "",
            discipline: o["discipline"] as? String ?? "BOULDER",
            geometry: o["geometry"] as? String ?? "POINT",
            direction: o["direction"] as? String ?? "LTR",
            sectorId: o["sectorId"] as? String,
            orientation: o["orientation"] as? String,
            path: o["path"] as? [[Double]] ?? [],
            faces: caras,
            savedAt: o["savedAt"] as? Double ?? 0
        )
    }

    static func clear(schoolId: String) {
        try? FileManager.default.removeItem(at: ficheroJson(schoolId))
        // Las fotos de esa escuela se van con él.
        let prefijo = seguro(schoolId) + "-cara"
        let todos = (try? FileManager.default.contentsOfDirectory(atPath: carpeta.path)) ?? []
        for f in todos where f.hasPrefix(prefijo) {
            try? FileManager.default.removeItem(at: carpeta.appendingPathComponent(f))
        }
    }

    /// ¿Hay algo que merezca la pena guardar? Un formulario en blanco no.
    static func tieneContenido(_ d: Draft) -> Bool {
        if !d.name.trimmingCharacters(in: .whitespaces).isEmpty { return true }
        return d.faces.contains { cara in
            cara.photo != nil || cara.blocks.contains {
                !$0.name.trimmingCharacters(in: .whitespaces).isEmpty
                    || $0.grade != nil || !$0.line.isEmpty
            }
        }
    }
}
